src/java/main/org/apache/zookeeper/common/ZKTrustManager.java

```diff
@@ -0,0 +1,137 @@
+        this.x509ExtendedTrustManager = x509ExtendedTrustManager;
+        this.hostnameVerificationEnabled = hostnameVerificationEnabled;
+        this.shouldVerifyClientHostname = shouldVerifyClientHostname;
+
+        hostnameVerifier = new DefaultHostnameVerifier();
+    }
+
+    @Override
+    public X509Certificate[] getAcceptedIssuers() {
+        return x509ExtendedTrustManager.getAcceptedIssuers();
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        if (hostnameVerificationEnabled && shouldVerifyClientHostname) {
+            performHostVerification(socket.getInetAddress(), chain[0]);
+        }
+        x509ExtendedTrustManager.checkClientTrusted(chain, authType, socket);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        if (hostnameVerificationEnabled) {
+            performHostVerification(socket.getInetAddress(), chain[0]);
+        }
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType, socket);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine) throws CertificateException {
+        if (hostnameVerificationEnabled && shouldVerifyClientHostname) {
+            try {
+                performHostVerification(InetAddress.getByName(engine.getPeerHost()), chain[0]);
+            } catch (UnknownHostException e) {
+                throw new CertificateException("failed to verify host", e);
+            }
+        }
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType, engine);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine engine) throws CertificateException {
+        if (hostnameVerificationEnabled) {
+            try {
+                performHostVerification(InetAddress.getByName(engine.getPeerHost()), chain[0]);
+            } catch (UnknownHostException e) {
+                throw new CertificateException("failed to verify host", e);
+            }
+        }
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType, engine);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        x509ExtendedTrustManager.checkClientTrusted(chain, authType);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType);
+    }
+
+    private void performHostVerification(InetAddress inetAddress, X509Certificate certificate) throws CertificateException {
+        try {
+            hostnameVerifier.verify(inetAddress.getHostAddress(), certificate);
```

Comment:
I agree. It could also be argued that authorization is handled by this patch in its current state as well, by only accepting peers that have trusted certificates. In other words, the CA limits which machines get certificates, which limits which machines get to join the quorum. So check 2 is not "needed" if the CA is trusted. 

I also think that check 2 has value for deployments that want to reuse the same truststore for multiple services and for deployments that are not using SSL/TLS at all. So I think it is a check worth doing and it is deserving of its own patch. Would be interested in knowing what other members of the community think.